Build failure: sharkey

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

added the following to my configuration on unstable:
```nix
services.sharkey = {
  enable = true;
  settings.url = "example.com";
  package = pkgs-unstable.sharkey;
};
```
This build failure also happened when using unstable for all packages and options, but I have since moved to stable by default in part to avoid this error, hence the `pkgs-unstable` prefix.


### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/317944547

### Relevant log output

```console
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/ahvg92hb6kdbjxv6pj12zzhlk8dqbqnr-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Executing pnpmConfigHook
Found 'pnpm' with version '9.15.9'
Using fetcherVersion: 2
Configuring pnpm store
/build /build/source
/build/source
Installing dependencies
[41m[30m ERR_PNPM_UNSUPPORTED_ENGINE [39m[49m [31mUnsupported environment (bad pnpm and/or Node.js version)[39m

Your Node version is incompatible with "/build/source/packages/backend".

Expected version: ^22.0.0
Got: v24.12.0

This is happening because the package's manifest has an engines.node field specified.
To fix this issue, install the required Node version.

ERROR: pnpm failed to install dependencies

If you see ERR_PNPM_NO_OFFLINE_TARBALL above this, follow these to fix the issue:
1. Set pnpmDeps.hash to "" (empty string)
2. Build the derivation and wait for it to fail with a hash mismatch
3. Copy the 'got: sha256-' value back into the pnpmDeps.hash field
```

### Additional context

I tried to (using a local copy of the nixpkgs repo) pin the nodejs version to nodejs_22 [as seen here](https://github.com/NixOS/nixpkgs/commit/268441e431da469252e086320108a75c283957b9), but this did not change the output. I'm not familiar with the source code, or nodejs/pnpm and so could not figure out a next step in troubleshooting.

### System metadata

fetching path input 'path:/nix/store/65lvd2sgh752p580mk6016wdkk3w828j-source'
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.63, NixOS, 25.11 (Xantusia), 25.11.20260103.30a3c51`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.93.3
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux, x86_64-v4-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/blltrx/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/blltrx/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/blltrx/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/blltrx/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/zp4ccr3fvndjx0dznjg7yf49xh68bdzw-lix-2.93.3/share`
 - channels(root): `"home-manager, nixos, nixos-unstable, nixpkgs"`
 - nixpkgs: `/nix/store/65lvd2sgh752p580mk6016wdkk3w828j-source`

### Notify maintainers

@hesiod 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/sh/sharkey/package.nix
+++ b/pkgs/by-name/sh/sharkey/package.nix
@@ -7,20 +7,25 @@
   cairo,
   cctools,
   ffmpeg-headless,
+  giflib,
   jemalloc,
   makeWrapper,
   nix-update-script,
-  nodejs,
+  nodejs_22,
   pango,
   pixman,
   pkg-config,
-  pnpm_9,
+  pnpm_10,
   fetchPnpmDeps,
   pnpmConfigHook,
   python3,
   vips,
   xcbuild,
 }:
+
+let
+  pnpm = pnpm_10.override { nodejs = nodejs_22; };
+in
 stdenv.mkDerivation (finalAttrs: {
   pname = "sharkey";
   version = "2025.4.4";
@@ -36,29 +41,32 @@ stdenv.mkDerivation (finalAttrs: {
 
   pnpmDeps = fetchPnpmDeps {
     inherit (finalAttrs) pname version src;
-    pnpm = pnpm_9;
-    fetcherVersion = 2;
-    hash = "sha256-34X8oJGkGXB9y7W4MquUkv8vY5yq2RoGIUCbjYppkIU=";
+    pnpm = pnpm;
+    fetcherVersion = 3;
+    hash = "sha256-wrA5Huv7b/P+5MNbScN9KzNwdHMtuceHu+Lw/C9lKlI=";
   };
 
   nativeBuildInputs = [
     makeWrapper
-    nodejs
+    nodejs_22
     pkg-config
     pnpmConfigHook
-    pnpm_9
+    pnpm
     python3
   ]
-  ++ (lib.optionals stdenv.hostPlatform.isDarwin [
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [
     cctools
     xcbuild
-  ]);
+  ];
 
   buildInputs = [
     cairo
     pango
     pixman
     vips
+  ]
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    giflib
   ];
 
   buildPhase = ''
@@ -71,7 +79,7 @@ stdenv.mkDerivation (finalAttrs: {
     popd
 
     # rebuild some node modules that have native dependencies
-    export npm_config_nodedir=${nodejs}
+    export npm_config_nodedir=${nodejs_22}
 
     pushd node_modules/.pnpm/node_modules/re2
     pnpm rebuild
@@ -82,12 +90,13 @@ stdenv.mkDerivation (finalAttrs: {
     popd
 
     pushd node_modules/.pnpm/node_modules/canvas
-    pnpm run install
+    NIX_CFLAGS_COMPILE="-include cstdint $NIX_CFLAGS_COMPILE" pnpm run install
     popd
 
     pnpm build
     node scripts/trim-deps.mjs
-    pnpm prune --prod --ignore-scripts
+    # FIXME: pruning is broken at the moment
+    # CI=true pnpm prune --prod --ignore-scripts
 
     runHook postBuild
   '';
@@ -96,8 +105,8 @@ stdenv.mkDerivation (finalAttrs: {
     let
       binPath = lib.makeBinPath [
         bash
-        nodejs
-        pnpm_9
+        nodejs_22
+        pnpm
       ];
       libPath = lib.makeLibraryPath [
         ffmpeg-headless
@@ -134,7 +143,7 @@ stdenv.mkDerivation (finalAttrs: {
       cp -r tossface-emojis/dist $out/sharkey/tossface-emojis/
 
       # create a wrapper script for running sharkey commands (ie. alias for pnpm run)
-      makeWrapper ${lib.getExe pnpm_9} $out/bin/sharkey \
+      makeWrapper ${lib.getExe pnpm} $out/bin/sharkey \
         --chdir $out/sharkey \
         --add-flags run \
         --set-default NODE_ENV production \

